sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp

```diff
@@ -4,18 +4,635 @@
+  std::string responseHeaders = WideStringToString(std::wstring(start, last - start));
+
+  DWORD sizeOfHttp = sizeOfHeaders;
+
+  // Get the HTTP version.
+  if (!WinHttpQueryHeaders(
+          handleManager->m_requestHandle,
+          WINHTTP_QUERY_VERSION,
+          WINHTTP_HEADER_NAME_BY_INDEX,
+          outputBuffer.data(),
+          &sizeOfHttp,
+          WINHTTP_NO_HEADER_INDEX))
+  {
+    GetErrorAndThrow("Error while querying response headers.");
+  }
+
+  start = outputBuffer.data();
+  // Assuming ASCII here is OK since the input is expected to be an HTTP version string.
+  std::string httpVersion = WideStringToStringASCII(start, start + sizeOfHttp / sizeof(WCHAR));
+
+  uint16_t majorVersion = 0;
+  uint16_t minorVersion = 0;
+  ParseHttpVersion(httpVersion, &majorVersion, &minorVersion);
+
+  DWORD statusCode = 0;
+  DWORD dwSize = sizeof(statusCode);
+
+  // Get the status code as a number.
+  if (!WinHttpQueryHeaders(
+          handleManager->m_requestHandle,
+          WINHTTP_QUERY_STATUS_CODE | WINHTTP_QUERY_FLAG_NUMBER,
+          WINHTTP_HEADER_NAME_BY_INDEX,
+          &statusCode,
+          &dwSize,
+          WINHTTP_NO_HEADER_INDEX))
+  {
+    GetErrorAndThrow("Error while querying response headers.");
+  }
+
+  HttpStatusCode httpStatusCode = static_cast<HttpStatusCode>(statusCode);
+
+  // Get the optional reason phrase.
+  std::string reasonPhrase;
+  DWORD sizeOfReasonPhrase = sizeOfHeaders;
+
+  if (WinHttpQueryHeaders(
+          handleManager->m_requestHandle,
+          WINHTTP_QUERY_STATUS_TEXT,
+          WINHTTP_HEADER_NAME_BY_INDEX,
+          outputBuffer.data(),
+          &sizeOfReasonPhrase,
+          WINHTTP_NO_HEADER_INDEX))
+  {
+    start = outputBuffer.data();
+    reasonPhrase
+        = WideStringToString(std::wstring(start, start + sizeOfReasonPhrase / sizeof(WCHAR)));
+  }
+
+  // Allocate the instance of the response on the heap with a shared ptr so this memory gets
+  // delegated outside the transport and will be eventually released.
+  auto rawResponse
+      = std::make_unique<RawResponse>(majorVersion, minorVersion, httpStatusCode, reasonPhrase);
+
+  AddHeaders(responseHeaders, rawResponse);
+
+  int64_t contentLength
+      = GetContentLength(handleManager, requestMethod, rawResponse->GetStatusCode());
+
+  rawResponse->SetBodyStream(std::move(GetBodyStream(handleManager, contentLength)));
+  return rawResponse;
+}
+
+std::unique_ptr<RawResponse> WinHttpTransport::Send(Context const& context, Request& request)
+{
+  // TODO: Need to test if context has been canceled and cleanup, can't call ThrowIfCanceled.
+  (void)(context);
+
+  auto handleManager = std::make_unique<Details::HandleManager>(context, request);
+
+  GetSessionHandle(handleManager);
+  GetConnectionHandle(handleManager);
+  GetRequestHandle(handleManager);
+
+  SendRequest(handleManager);
+
+  ReceiveResponse(handleManager);
+
+  return GetRawResponse(handleManager, request.GetMethod());
```

Comment:
All these methods dont's need to get a unique_ptr<handleManager>.
Instead, they can just take a referece or a raw pointer to HandleManager.

The benefit, if it just a reference, is that you never worry about null argument exception.
See, I made this example on how to pass a reference from a unique_ptr, I recommend following this approach here
https://godbolt.org/z/71G7nd

Reply:
Given this is an implementation detail and seems somewhat subjective, I am inclined to leave it as is for now.

One thing I need is the destructor of `handleManager` to be called whenever we throw. Would passing them as raw pointers or a reference still let that happen?